# STAT 207 Group Lab Assignment 13 - [10 total points]

## Measuring Model Performance & Model Searching

<hr>

## <u>Lab Grading</u>:

Should we grade your submission?  If not, write the netID of the submission to be graded.  (Note: We will only grade one assignment per group, and we'll pick the first one that says we should grade that submission.  We will assign the same grade to all team members.)

*For example*, you might respond: **grade this submission** or **my submission is under netID jdeeke**

If you said **my submission is under netID** above, we will not read any more of your lab submission.

If you said **grade this submission** above, who worked with you on this submission?  Write both their **names** and **netIDs**.  Also, discuss and record if you've ever been on a cruise before.

grade this submission 


amarg3 

## <u>Purpose</u>:
You should work in groups of 2-3 on this report (not working in groups without permission will result in a point deduction). The purpose of this group lab assignment is to assess model fit and overfitting through model searching.
<hr>

### Group Roles

Suggested and specified roles are provided below: 

#### Groups of 2

* **Driver**: This student will type the report.  While typing the report, you may be the one who is selecting the functions to apply to the data.
* **Navigator**: This student will guide the process of answering the question.  Specific ways to help may include: outlining the general steps needed to solve a question (providing the overview), locating examples within the course notes, and reviewing each line of code as it is typed.

#### Groups of 3

* **Driver**: This student will type the report.  They may also be the one to select the functions to apply to the data.
* **Navigator**: This student will guide the process of answering the question.  They may select the general approach to answering the question and/or a few steps to be completed along the way. 
* **Communicator**: This student will review the report (as it is typed) to ensure that it is clear and concise.  This student may also locate relevant examples within the course notes that may help complete the assignment.

<hr>

### Imports

In [48]:
#Run this
import pandas as pd                    # imports pandas and calls the imported version 'pd'
import matplotlib.pyplot as plt        # imports the package and calls it 'plt'
import seaborn as sns                  # imports the seaborn package with the imported name 'sns'
sns.set()  
from sklearn.linear_model import LinearRegression
import statsmodels.formula.api as smf
from sklearn.model_selection import train_test_split

## Case Study: Crew Size on Cruise Ships

We will look at data collected about the size of a crew (in hundreds) on cruise ships.  In particular, we would like to be able to predict the size of the crew needed for a cruise ship from other characteristics about that ship.  The original data can be found here: https://github.com/bot13956/ML_Model_for_Predicting_Ships_Crew_Size/blob/master/cruise_ship_info.csv.

For this case study, we will consider the following variables:

- **Age**: the age of the ship (years)
- **Tonnage**: the amount of water the ship displaces
- **passengers**: the max number of passengers the ship can hold (in 100s)
- **length**: the length of the ship (in hundrds of feet)
- **cabins**: the number of cabins (in 100s)
- **passenger_density**: a space ratio for how much space is on board per passenger
- **crew**: the number of crew members on the ship (in 100s)

The code cell below will read in the data for you and create a test and training set.  Be sure to run the cell. 

In [49]:
df = pd.read_csv('cruise_ship_info.csv')
df_train, df_test = train_test_split(df, test_size = 0.2, random_state = 425)
df

,Ship_name,Cruise_line,Age,Tonnage,passengers,length,cabins,passenger_density,crew
0,Journey,Azamara,6,30.277,6.94,5.94,3.55,42.64,3.55
1,Quest,Azamara,6,30.277,6.94,5.94,3.55,42.64,3.55
2,Celebration,Carnival,26,47.262,14.86,7.22,7.43,31.80,6.70
3,Conquest,Carnival,11,110.000,29.74,9.53,14.88,36.99,19.10
4,Destiny,Carnival,17,101.353,26.42,8.92,13.21,38.36,10.00
...,...,...,...,...,...,...,...,...,...
153,Taurus,Star,22,3.341,0.66,2.79,0.33,50.62,0.59
154,Virgo,Star,14,76.800,19.60,8.79,9.67,39.18,12.00
155,Spirit,Windstar,25,5.350,1.58,4.40,0.74,33.86,0.88
156,Star,Windstar,27,5.350,1.67,4.40,0.74,32.04,0.88


### 1. [4 points] Searching for a Good Model

**a)** Perform forward selection using the adjusted $R^2$ ($R^2_\text{adj}$) as your metric on your training data.

In [50]:
current_best_model = smf.ols(formula='crew~1', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(-2.220446049250313e-16)

In [51]:
current_best_model = smf.ols(formula='crew~Age', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.3419120323647763)

In [52]:
current_best_model = smf.ols(formula='crew~Tonnage', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.8678289039373271)

In [53]:
current_best_model = smf.ols(formula='crew~passengers', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.8531635824427954)

In [54]:
current_best_model = smf.ols(formula='crew~length', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.7975697196171544)

In [55]:
current_best_model = smf.ols(formula='crew~cabins', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.9201134337352352)

In [56]:
current_best_model = smf.ols(formula='crew~passenger_density', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(-0.0025789914438698602)

In [57]:
current_best_model = smf.ols(formula='crew~cabins+Age', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.9207476460815851)

In [58]:
current_best_model = smf.ols(formula='crew~cabins+Tonnage', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.9233088184185861)

In [59]:
current_best_model = smf.ols(formula='crew~cabins+passengers', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.9222580682513719)

In [60]:
current_best_model = smf.ols(formula='crew~cabins+length', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.927124418689438)

In [61]:
current_best_model = smf.ols(formula='crew~cabins+passenger_density', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.9253408973356728)

In [62]:
current_best_model = smf.ols(formula='crew~cabins+length+Age', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.9269407967317781)

In [63]:
current_best_model = smf.ols(formula='crew~cabins+length+Tonnage', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.926854849890568)

In [64]:
current_best_model = smf.ols(formula='crew~cabins+length+passengers', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.9310097480333617)

In [65]:
current_best_model = smf.ols(formula='crew~cabins+length+Tonnage+passenger_density', data=df_train).fit()
current_best_model.rsquared_adj

np.float64(0.928793034526106)

**b)** Briefly summarize the results of this selection process.  Be sure to answer:
- What was your starting model?
- What was the most important initial variable to add to the model?
- What was your final model?
- What was the $R^2_\text{adj}$ of your final model?

Our starting model was crew size because that is what we were trying to predict. Our most important initial value was cabins because it had the largest r^2 and final model was cabins+length+passengers and the r^2 adjusted value was .931.

### 2. [2.5 points]  Check for Multicollinearity

**a)** Now, check your model that you selected in Question 1 after forward selection for multicollinearity.  You may use either a visualization or summary statistics to check.

In [66]:
df[["crew", "cabins", "passengers", "length"]].corr()

,crew,cabins,passengers,length
crew,1.000000,0.950823,0.915234,0.895857
cabins,0.950823,1.000000,0.976341,0.889798
passengers,0.915234,0.976341,1.000000,0.883535
length,0.895857,0.889798,0.883535,1.000000


**b)** Based on what you see from part a, which variables (if any) would you like to remove from your model to reduce multicollinearity?  Briefly explain how you selected your variables that you wanted to remove.  Then, fit a new model to the training data using your reduced set of selected predictors.

*Note*: If you are satisfied with your current model for multicollinearity, pick the least important predictor to remove from the model, and fit an updated model to the training data with your selected predictors.  Be sure to explain why you did not want to remove any variables.

*Second note*: there is not a single correct answer here.  We are hoping that you make a reasonable choice based on your current understanding of the data.

There is a lot of multicolinearty among many predictors. Cabin and passangers has a really high correlation. To reduce multicolinearity, we would remove passenger leaving only cabin and length as our predictors because cabin was very important during our forward selection. 

In [68]:
current_model = smf.ols(formula = "crew~cabins+length", data=df_train).fit()
current_model.rsquared_adj

np.float64(0.927124418689438)

### 3. [3.5 points] Comparing Our Two Models for Overfitting

We now have two candidate models: our model from forward selection in Question 1 and our reduced model from Question 2.  In this question, we'll aim to identify which model performs better on *new* data.

**a)** Calculate the RMSE for each of our two models on each of our two data sets: the training and the test set.  Be sure that you use the same original model for both calculations (that is, you don't need to re-fit a second version of your model to your test data).

In [69]:
full_model = smf.ols(formula="crew~cabins+length+passengers", data=df_train).fit()
full_model.rsquared_adj

np.float64(0.9310097480333617)

In [70]:
reduced_model = smf.ols(formula="crew~cabins+length", data=df_train).fit()
reduced_model.rsquared_adj

np.float64(0.927124418689438)

In [72]:
from sklearn.metrics import mean_squared_error
import numpy as np

full_train_pred = full_model.predict(df_train)
full_test_pred = full_model.predict(df_test)

full_train_rmse = np.sqrt(mean_squared_error(df_train["crew"], full_train_pred))
full_test_rmse = np.sqrt(mean_squared_error(df_test["crew"], full_test_pred))

full_train_rmse, full_test_rmse

(np.float64(0.9216190394055853), np.float64(1.1874445351046277))

In [73]:
full_train_pred = reduced_model.predict(df_train)
full_test_pred = reduced_model.predict(df_test)

full_train_rmse = np.sqrt(mean_squared_error(df_train["crew"], full_train_pred))
full_test_rmse = np.sqrt(mean_squared_error(df_test["crew"], full_test_pred))

full_train_rmse, full_test_rmse

(np.float64(0.9510891536977195), np.float64(1.2413414311152016))

**b)** Based on these results, which model would you prefer?  Briefly explain how you picked your preferred model.

I prefer the full model because the test RMSE for the full model (1.18) was lower than the reduced (1.24).

**c)** Do your results from Question 1 (forward searching) and Question 3 (RMSE) suggest using the same model?  

We used three different metrics for our three different questions: $R^2_\text{adj}$ (Question 1), multicollinearity (Question 2), and RMSE (Question 3).  These metrics prioritize different characteristics of a model.  What characteristic of a model do these metrics prioritize?

*Note:* You should answer this question separately for each metric.

Yes, the results from question 1 and 3 suggest using the same model since the adjusted r^2 focuses on blanacing fit and finding a number of predictors. Multicollinearity prioritizes stable predictors and variables.

### 4. AI Acknowledgement

Our course policy is that you should write all of your own interpretations and other narrative answers (phrases or sentences) yourself without the assistance of AI.  AI should not be used on Questions 1b, 2b, 3b, and 3c of this assignment.  

You may use AI to help guide your code, although you should write all of your own code yourself (not copy-paste from another source) and you should cite your use of AI.  I would encourage you to try to generate any necessary code yourself first using course resources and using AI as a debugging tool if/when you reach an error that you can't figure out or to help you perform any coding tasks that are more advanced than we've demonstrated during class (intended only for projects).  

Did you use AI on this assignment?  Did you use other resources outside of our course-provided resources on this assignment?

no

If you used AI or other resources, answer the following questions to cite your usage.

- Which AI and/or resources did you use?
- What prompts did you ask it?
- How did you integrate the responses into your assignment?  Specifically, which questions or parts are associated with this usage?

Note: answering these three questions are enough for our course but may not be enough for a different course or context.